sky/shell/tracing_controller.cc

```diff
@@ -19,11 +20,85 @@ namespace shell {
 const char kBaseTraceStart[] = "{\"traceEvents\":[";
 const char kBaseTraceEnd[] = "]}";
 
+const char kObservatoryMethodStartTracing[] = "flutter_startTracing";
+const char kObservatoryMethodStopTracing[] = "flutter_stopTracing";
+const char kObservatoryResultOk[] = "{\"success\" : true}";
+const char kObservatoryResultFail[] = "{\"success\" : false}";
+
+static const char* ObservatoryInvoke(const char* method,
+                                     const char** param_keys,
+                                     const char** param_values,
+                                     intptr_t num_params,
+                                     void* user_data) {
+  if (user_data == nullptr) {
+    return strdup(kObservatoryResultFail);
+  }
+
+  auto tracingController = reinterpret_cast<TracingController*>(user_data);
+
+  if (strncmp(method, kObservatoryMethodStartTracing,
+              sizeof(kObservatoryMethodStartTracing)) == 0) {
+    tracingController->StartTracing();
+    return strdup(kObservatoryResultOk);
+  }
+
+  if (strncmp(method, kObservatoryMethodStopTracing,
+              sizeof(kObservatoryMethodStopTracing)) == 0) {
+    // Flushing the trace log requires an active message loop. However,
+    // observatory callbacks are made on a dart worker thread. We setup a
+    // message loop manually and tell the flush completion handler to terminate
+    // the loop when done
+    base::MessageLoop workerThreadLoop;
+
+    base::FilePath tempDir;
+    bool tempAccess = base::GetTempDir(&tempDir);
```

Comment:
`base::GetTempDir` does not create the temporary directory. Only returns the path to the platform temp directory. Likewise, using FLAG_OPEN and FLAG_READ will not create the file if it does not exist. If it does not, the fstat fails and returns size 0. AFAIK, all file handles are cleaned up and temp files collected.